tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc

```diff
@@ -176,6 +176,297 @@ struct ConcatOpConverter : public OpRewritePattern<ConcatenateOp> {
+    return rewriter.create<mlir::ConstantIntOp>(loc, 0, intType.getWidth());
+
+  FloatType floatType = eltType.cast<FloatType>();
+  return rewriter.create<mlir::ConstantFloatOp>(
+      loc, APFloat::getZero(floatType.getFloatSemantics()), floatType);
+}
+
+/// Emits a nest to fill the given `buffer` of memref type with fillValue.
+static LogicalResult fillBuffer(Location loc, Value buffer, Value fillValue,
+                                PatternRewriter& builder) {
+  OpBuilder::InsertionGuard guard(builder);
+  MemRefType bufType = buffer.getType().cast<MemRefType>();
+  unsigned rank = bufType.getRank();
+  SmallVector<Value, 4> dimSizes;
+  dimSizes.reserve(rank);
+  for (unsigned i = 0; i < rank; ++i)
+    dimSizes.push_back(builder.create<mlir::memref::DimOp>(loc, buffer, i));
+
+  AffineMap idSymMap = builder.getSymbolIdentityMap();
+  AffineMap lbMap = builder.getConstantAffineMap(0);
+  SmallVector<Value, 4> ivs(rank);
+  AffineForOp forOp;
+  for (unsigned i = 0; i < rank; ++i) {
+    forOp = builder.create<AffineForOp>(loc, llvm::None, lbMap, dimSizes[i],
+                                        idSymMap);
+    builder.setInsertionPointToStart(forOp.getBody());
+    ivs[i] = forOp.getInductionVar();
+  }
+  Type fillValueType = fillValue.getType();
+  auto fillMemRefType = fillValueType.dyn_cast<MemRefType>();
+  assert(((fillMemRefType && fillMemRefType.getRank() == 0) ||
+          fillValueType.isIntOrFloat()) &&
+         "init value has to be a 0-d memref or int or fp");
+  Value initVal = fillMemRefType
+                      ? builder.create<AffineLoadOp>(loc, fillValue, llvm::None)
+                      : fillValue;
+  builder.create<AffineStoreOp>(loc, initVal, buffer, ivs);
+  return success();
+}
+
+class GatherOpConverter : public OpRewritePattern<GatherOp> {
+ public:
+  using OpRewritePattern<GatherOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(GatherOp op,
+                                PatternRewriter& rewriter) const final {
+    Location loc = op.getLoc();
+
+    // Operand array.
+    Value operand = op.operand();
+    MemRefType operand_type = operand.getType().cast<MemRefType>();
+    unsigned operand_rank = operand_type.getRank();
+    ArrayRef<int64_t> operand_shape = operand_type.getShape();
+
+    // Start_indices array.
+    Value start_indices = op.start_indices();
+    MemRefType start_indices_type = start_indices.getType().cast<MemRefType>();
+    unsigned start_indices_rank = start_indices_type.getRank();
+    ArrayRef<int64_t> start_indices_shape = start_indices_type.getShape();
+
+    // Output array.
+    Value output = op.output();
+    MemRefType output_type = output.getType().cast<MemRefType>();
+    unsigned output_rank = output_type.getRank();
+    ArrayRef<int64_t> output_shape = output_type.getShape();
+
+    if (!operand_type.hasStaticShape() ||
+        !start_indices_type.hasStaticShape() || !output_type.hasStaticShape())
+      return failure();
```

Comment:
Use `notifyMatchFailure("...")` - this will provide a useful message for debugging.